### DISCUSSION TITLE
Very slow during tests

### DISCUSSION THREAD

**Discussion by dargue3** [mswjs/msw#659] [Q&A]
Hey! I'm sure this is going to be difficult for anyone to debug without a repro, but just taking a shot in the dark.

My jest tests are running SUPER slow and when I use `console.time()` to see how long it's taking for MSW to return some mocked responses, I'm seeing upwards of 3 seconds on a few of the graphql queries for the MSW resolver to even be called, and then another 3 for the response to return back to the caller. When I perform the same action in the browser — 50ms.

I'm assuming MSW can only respond as fast as the CPU has time for, which leads me to believe that maybe Jest is just hogging resources so much MSW can't even find a spare tick to do its response?

**Comment by msutkowski** [comment]
@dargue3 Is this the same with a `rest` request, or specifically limited to `graphl` operations? We can create a very simple repro and see if there are any stark differences.

**Reply by dargue3** [reply]
It's the same for either, though it does seem to scale based on the complexity of the test. But for sure, even the most basic test my app has, when I run that flow in Jest my `console.time()` reports 250ms, in the browser, 19ms. I'll try to create something really basic... Do you happen to have any snowpack + jest setups I could clone?

**Reply by msutkowski** [reply]
There is an example with jest/graphql here: https://github.com/mswjs/examples/tree/master/examples/graphql-react-apollo. I do not have one with snowpack, but would be happy to help out with creating a specific repro. In the linked example repo, that simple test executes in under 20ms on my machine.

Does running your tests with `--runInBand` make any difference?

**Reply by dargue3** [reply]
`runInBand` seems to help when it's _one_ test, but of course once you're running several of them it's definitely a much slower option.

**Reply by msutkowski** [reply]
> `runInBand` seems to help when it's _one_ test, but of course once you're running several of them it's definitely a much slower option.

Unfortunately, at this time, we always recommend running with `--runInBand` enabled. More info: https://github.com/mswjs/msw/issues/474

**Reply by dargue3** [reply]
oh interesting. Okay, I'll keep this in mind. But for the record, this wouldn't really have anything to do with how fast MSW seems to be responding, correct? I would assume this is more on the side of Jest just completely bottlenecking the CPU during tests

**Reply by dargue3** [reply]
I tried out the basic repro and (of course haha) Jest and browser running exactly the same speed. Going to try to add more pieces of my puzzle to the min repro and see if anything breaks it

**Reply by msutkowski** [reply]
> oh interesting. Okay, I'll keep this in mind. But for the record, this wouldn't really have anything to do with how fast MSW seems to be responding, correct? I would assume this is more on the side of Jest just completely bottlenecking the CPU during tests

Not that I'm aware of, no. Jest _should_ run [one test per CPU](https://jestjs.io/docs/cli#--maxworkersnumstring) in it's normal parallel mode. You should try debugging with [--logHeapUsage](https://jestjs.io/docs/cli#--logheapusage) and see if there is anything going as you add more of your stuff to it.

**Reply by dargue3** [reply]
Okay I've created somewhat of a repro [here](https://github.com/dargue3/modified-msw-examples). Follow the readme for installing as normal and then from within `examples/graphql-react-apollo` run `yarn test:unit`. You should see a couple `console.time` print outs. For me, the `GetUser` clocks in at about ~45ms but when I run in the browser, 14ms.

It's not quite 3 seconds, but I think the amount of time scales up with the amount of complexity in the app. My codebase has about 95 integration tests and 45 MSW handlers.

**Reply by msutkowski** [reply]
> Okay I've created somewhat of a repro [here](https://github.com/dargue3/modified-msw-examples). Follow the readme for installing as normal and then from within `examples/graphql-react-apollo` run `yarn test:unit`. You should see a couple `console.time` print outs. For me, the `GetUser` clocks in at about ~45ms but when I run in the browser, 14ms.
> 
> It's not quite 3 seconds, but I think the amount of time scales up with the amount of complexity in the app. My codebase has about 95 integration tests and 45 MSW handlers.

Sorry about the delay - I'll give this a look in the morning if someone doesn't beat me to it :)

**Reply by kettanaito** [reply]
Note that you can run your tests with the `DEBUG` environmental variable and observe the internals of the Node.js interception. Those logs also include timestamps and time diffs, making it easier to see which step of request handling takes more time than necessary. 

```sh
DEBUG=* npm test
```

**Reply by dargue3** [reply]
I tried that once and the output was so massive it was hard for me to parse. My tests render entire pages at a time so there is often like 10 mocked endpoints all firing. 

Honestly, I've switched my heavy duty tests over to Cypress which is performing well for me. I think Jest and jsdom just bog down way too heavily under this sort of load.

**Reply by msutkowski** [reply]
I did run your repro along with the `DEBUG` flag. I'm going to need to create a better testing environment, as it seems like the timers are out of sync. The XHR patching from DEBUG is saying it takes +/- 300ms in total, but the tests execute in about 14ms on my machine. 

I'd be really surprised if Cypress was able to run things much faster, and if it is, I'd like to understand why.

**Reply by dargue3** [reply]
I've tried a lot of things to measure performance, the best thing I could really manage to do is just litter `console.time()` both in the tests and within the code and just measure how long things take. I had one test that would take about 26 seconds to do something that cypress can do in 1500ms. 

Profiling CPU in jest tests is quite hard, flame charts don't really look like much without knowing exactly how all the internals of React works.

**Reply by kettanaito** [reply]
> I tried that once and the output was so massive it was hard for me to parse.

You can scope the output by providing the name of the logger. If you're running MSW in Node.js (via `setupServer`), you may scope the logs be doing either:

```sh
DEBUG=XHR* npm test
DEBUG=http* npm test
```

Those will depend on what kind of module performs the requests (http/https or XMLHttpRequest).

@dargue3, how big is your test suite? Can you reproduce the issue in isolation (i.e. a new repo, minimal app setup)? Does this performance decrease happen to all the tests or only some of them? Can you spot some patterns as to when it happens?

**Reply by dargue3** [reply]
I'll work on it, but honestly, the more that I dig into this, the more I think I can just attribute it to `jsdom` not performing well with large codebases. 

I have about 120 components on just 4 separate "pages", and in my tests I render a whole page and test the workflows. Basically none of my individual tests run quicker than 7 seconds or so purely do to the time it takes to render. I think whatever slowness I'm seeing with MSW can be attributed to jest + jsdom tanking the CPU. Run the same tests in the browser (like with Cypress) and everything is much much snappier. MSW shines with cypress!

**Reply by kettanaito** [reply]
> MSW shines with cypress!

Feedback like this warms my heart. 

Still, we cannot toss aside the possibility that there may be some issue on our end, so we'd appreciate your investigation.

**Comment by kettanaito** [comment]
There has been a memory leak issue reported and fixed in the latest release (0.29.0). There's a serious change it might have caused this behavior. It's highly recommended to update to the latest version.
